### ISSUE TITLE
AV while scrolling with scrollbuttons in themed app

### ISSUE THREAD

**Issue by bosalb** [JAM-Software/Virtual-TreeView#1170]
In version 7.6.3 a new bug was added, if you move the vertical scrollbar with the buttons in a themed app and at the same time the horizontal scrollbar is visible. This leads to an AV.

**Comment by joachimmarder** [comment]
Please respect our guidelines on the project homepage for submitting bugs. Please include your version of Delphi, and attach a sample compiling project as ZIP to your report, or instructions how to modify one of the included sample projects to reproduce the problem. 

A call stack of the exception would also be helpful.

I am unable to replicate the exception in the latest Delphi version.

**Comment by bosalb** [comment]
Sorry about the missing project. But i have to extract it from our large project.
I have attached a sample project for RAD Studio 11.2. It is written in C++, but you can reproduce it in Delphi too.
It is important to know, that the app must be themed. We use the "Windows 11 Modern style". 
Start the app , scroll with the vertical trackbar to bottom and you we see, that the horizontal scrollbar is not be themed (see After_Start.png). If you click into the TVirtualStringTree control, the scrollbar will be themed correctly.
If you now press the up button of the vertical scrollbar so often, until the horizontal scrollbar has gone, you get the AV: "Range override for line index". A screen of stacktrace is attached.
![After_Select_item](https://user-images.githubusercontent.com/125001627/225341828-ac59d532-b859-4aa7-9e0f-f4e1813e572f.png)
![After_Start](https://user-images.githubusercontent.com/125001627/225341835-75640e47-ff21-4e9d-a650-fb63fdf26dbb.png)
![Callstack](https://user-images.githubusercontent.com/125001627/225341837-998d79a1-6843-4071-898a-b656af75acaf.png)
[VirtualStringTree.zip](https://github.com/JAM-Software/Virtual-TreeView/files/10981037/VirtualStringTree.zip)

**Comment by joachimmarder** [comment]
>  It is written in C++

Sorry, I don't have C++ Builder. 

> scroll with the vertical trackbar to bottom and you we see, that the horizontal scrollbar is not be themed

I was unable to replicate this with the included "Advanced" sample project under Delphi 11.3.
Can you please try if the latest version 7.6.4 of Virtual TreeView changes anything for you?
This was the version I did the tests with.

**Comment by zencode1** [comment]
In RAD Studio 11.3, I built the project using the current V7_stable branch instead of the source included in the .zip (there were some missing files).  I don't have "Windows 11 Modern Style" so I chose "Windows 10".  I was able to see the horizontal scrollbar draw without theming and then switch to being drawn correctly using the style.  However, I was unable to reproduce the exception.

**Comment by bosalb** [comment]
Pleace install "Windows 11 Modern light" Style. You can get it from GetIt Package manager for free (see attached screen).  In "Windows 10" style, the problem is not reproducable. I test it again with 7.6.4 sources and it is still reproduceable.
![Windows 11 modern style](https://user-images.githubusercontent.com/125001627/225960289-9be2768d-b60b-4e1e-a478-8fadf760ff46.png)

**Comment by zencode1** [comment]
Ok, I installed that style and can confirm it is happening for me as well.  I tested in 32 and 64 bit to be sure.

**Comment by zencode1** [comment]
This happens when `TVclStyleScrollBarsHook.DrawHorzScrollBar` draws the horizontal slider and HorzScrollRect has no width or height.  I don't know the correct fix, but a workaround for the exception is to check HorzScrollRect and exit if it's not right.
```
procedure TVclStyleScrollBarsHook.DrawHorzScrollBar(DC: HDC);
...
    B := TBitmap.Create;
    try
      R := HorzScrollRect;
      if (R.Width = 0) or (R.Height = 0) then
        Exit;
...
```
or, do it before the bitmap is created:
```
    R := HorzScrollRect;
    if (R.Width = 0) or (R.Height = 0) then
      Exit;
    B := TBitmap.Create;
    try
...
```

**Comment by joachimmarder** [comment]
`VclStyleScrollBarsHook.DrawHorzScrollBar()`should only be used in RAD Studio 10.3 and below. But this issue refers to 11.2. Can you please check this again?

**Comment by zencode1** [comment]
@joachimmarder I'm debugging this test using Delphi on V7_stable.  From the files included in the .zip it appears @bosalb is using the same branch.

I see the `CompilerVersion < 34` in VirtualTrees.StyleHooks.pas on the `master` branch but there is no similar conditional check in `V7_stable`.  FWIW, I updated my test application to use `master` and it does not exhibit the behavior mentioned in this issue.  The theme draws correctly and there is no exception.

**Comment by joachimmarder** [comment]
I'm sorry, I mixed up the branches. 

In that case I will merge the unit `VirtualTrees.StyleHooks` from master to "v7_stable" branch. I originally thought the change might be too big for a bugfix release, but it seems to make sense then.

**Comment by joachimmarder** [comment]
Merged changes for #1016 to `v7_stable` branch.
